text/0046-metadata-for-offline-signers.md

```diff
@@ -0,0 +1,274 @@
+  spec_name: Vec<u8>, // SCALE from `String`
+  base58_prefix: u16,
+  decimals: u8,
+  token_symbol: Vec<u8>, // SCALE from `String`
+}
+```
+
+constitute metadata descriptor. This is minimal information that is, together with (shortened) types registry, sufficient to decode any signable transaction.
+
+#### Merkle tree
+
+A **Complete Binary Merkle Tree** (**CBMT**) is proposed as digest structure.
+
+Every node of the proposed tree has a 32-bit value.
+
+A terminal node of the tree we call **leaf**. Its value is input for digest.
+
+The top node of the tree we call **root**.
+
+All node values for non-leave nodes are not terminal are computed through non-commutative **merge** procedure of child nodes.
+
+In CBMT, all layers must be populated, except for the last one, that must have complete filling from the left.
+
+Nodes are numbered top-down and left-to-right starting with 0 at the top of tree.
+
+```
+Example 8-node tree
+
+        0
+     /     \
+    1       2
+   / \     / \
+  3   4   5   6
+ / \
+7   8
+
+Nodes 4, 5, 6, 7, 8 are leaves
+Node 0 is root
+
+```
+
+### General flow
+
+1. The metadata is converted into lean modular form (vector of chunks)
+2. A Merkle tree is constructed from the metadata chunks
+3. A root of tree (as a left element) is merged with Metadata Descriptor (as a right element)
+4. Resulting value is a constant to be included in `additionalSigned` to prove that the metadata seen by cold device is genuine
+
+### Metadata modularization
+
+1. Types registry is stripped from `docs` fields.
+2. Types records are separated into chunks, with enum variants being individual chunks differing by variant index; each chunk consisting of `id` (same as in full metadata registry) and SCALE-encoded 'Type' description (reduced to 1-variant enum for enum variants). Enums with 0 variants are treated as regular types.
+3. Chunks are sorted by `id` in ascending order; chunks with same `id` are sorted by enum variant index in ascending order.
+
+```
+types_registry = metadataV14.types
+modularized_registry = EmptyVector<id, type>
+for (id, type) in types.registry.iterate_enumerate {
+  type.doc = Null
+  if (type is ReduceableEnum) { // false for 0-variant enums
+    for variant in type.variants.iterate {
+      variant_type = Type {
+        path: type.path
+        type_params: Null
+        type_def: TypeDef::Variant(variants: [variant])
+      }
+      modularized_registry.push(id, variant_type)
+    }
+  } else {
+    modularized_registry.push(id, type)
+  }
+}
+
+modularized_registry.sort(|a, b| {
+    if a.id == b.id { //only possible for variants
+      a.variant_index > b.variant_index
+    } else { a.id > b.id }
+  }
+)
+
+```
+
+### Merging protocol
+
+`blake3` transformation of concatenated child nodes (`blake3(left + right)`) as merge procedure;
+
+### Complete Binary Merkle Tree construction protocol
+
+1. Leaves are numbered in ascending order. Leaf index is associated with corresponding chunk.
```

Comment:
Isn't it the other way around, you associate chunks with leaf indices? 

In general I would also by intuition, order all leaves in ascending order. Then I would chunk them into two and hash them. The output of this would follow the same procedure until only one hash (the root hash) is left.

Reply:
There are other approaches; this is kind of canonical for CBMT, but there could be implementers that do not know canonical stuff. I know we've independently made exactly same implementation for hot-to-cold proof transfer protocol with @carlosala based on natural order of universe, but I'm afraid here we should be a bit more explicit as canonical here is not the best construction path, just a convenient one. Ordering matters here, and also there is an issue of pairing last odd leaf that could be resolved in more awkward or even slightly less efficient way (making tree incomplete, but I'm not sure it's obvious for everyone).

Also, just following my description verbatim, it's possible to construct the tree without knowing much of background. That should just make implementations easier hopefully. I was thinking about also writing pseudocode representation, but that's probably too much already.